nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/excel/RowIterator.java

```diff
@@ -0,0 +1,155 @@
+                .collect(Collectors.toMap(key -> key, value -> Boolean.FALSE)) : new HashMap<>();
+        this.firstRow = firstRow;
+        this.logger = logger;
+        this.log = logger != null;
+    }
+
+    @Override
+    public boolean hasNext() {
+        setCurrent();
+        boolean next = currentRow != null;
+        if(!next) {
+            String sheetsNotFound = getSheetsNotFound(desiredSheets);
+            if (!sheetsNotFound.isEmpty() && log) {
+                logger.warn("Excel sheet(s) not found: {}", sheetsNotFound);
+            }
+        }
+        return next;
+    }
+
+    private void setCurrent() {
+        currentRow = getNextRow();
+        if (currentRow != null) {
+            return;
+        }
+
+        currentSheet = null;
+        currentRows = null;
+        while (sheets.hasNext()) {
+            currentSheet = sheets.next();
+            if (isIterateOverAllSheets() || hasSheet(currentSheet.getSheetName())) {
+                currentRows = currentSheet.iterator();
+                currentRow = getNextRow();
+                if (currentRow != null) {
+                    return;
+                }
+            }
+        }
+    }
+
+    private Row getNextRow() {
+        while (currentRows != null && !hasExhaustedRows()) {
+            Row tempCurrentRow = currentRows.next();
+            if (!isSkip(tempCurrentRow)) {
+                return tempCurrentRow;
+            }
+        }
+        return null;
+    }
+
+    private boolean hasExhaustedRows() {
+        boolean exhausted = !currentRows.hasNext();
+        if (log && exhausted) {
+            logger.info("Exhausted all rows from sheet {}", currentSheet.getSheetName());
+        }
+        return exhausted;
+    }
+
+    private boolean isSkip(Row row) {
+        return row.getRowNum() < firstRow;
+    }
+
+    private boolean isIterateOverAllSheets() {
+        boolean iterateAllSheets = desiredSheets.isEmpty();
+        if (iterateAllSheets && log) {
+            logger.info("Advanced to sheet {}", currentSheet.getSheetName());
+        }
+        return iterateAllSheets;
+    }
+
+    private boolean hasSheet(String name) {
+        boolean sheetByName = !desiredSheets.isEmpty()
+                && desiredSheets.keySet().stream()
+                .anyMatch(desiredSheet -> desiredSheet.equalsIgnoreCase(name));
+        if (sheetByName) {
+            desiredSheets.put(name, Boolean.TRUE);
+        }
+        return sheetByName;
+    }
+
+    private String getSheetsNotFound(Map<String, Boolean> desiredSheets) {
+        return desiredSheets.entrySet().stream()
+                .filter(entry -> !entry.getValue())
+                .map(Map.Entry::getKey)
+                .collect(Collectors.joining(","));
+    }
+
+    @Override
+    public Row next() {
+        return currentRow;
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.workbook.close();
+    }
+
+    void setLogger(ComponentLog logger) {
```

Comment:
As elsewhere, having a setter for the ComponentLog does not seem like the best approach, if the ComponentLog is needed, it should be passed in the constructor.

Reply:
Instead of passing a ComponentLog, could I just use a static final instance of `org.sl4j.Logger` to handle all logging?

Reply:
`ComponentLog` is necessary to associate log messages with particular instances of the component, so it is required. It should be possible to pass through the constructor.

Reply:
I guess my problem here is `RowIterator` is used by both `ExcelRecordReader` and `ExcelRecordSource`. Only `ExcelRecordReader` has an instance of `ComponentLog` while  `ExcelRecordSource` does not. That is why I have conditional logging depending on which class instantiates an instance of  `RowIterator`. I did not see how `ExcelRecordSource` could get an instance of `ComponentLog`. Please advise.

Reply:
`ExcelRecordReader` can pass its `ComponentLog` when instantiating `ExcelRecordSource`. Are there other places where `ComponentLog` is not available?

Reply:
Sounds good!